### ISSUE TITLE
Loss definition: from the paper to the implementation 

### ISSUE THREAD

**Issue by ReubenDo** [meng-tang/rloss#14]
Hello,

First I would like to say that I really enjoyed your work. Congrats!
I have a simple question regarding the denseCRF loss. I do not understand how your implementation works. Why is it a simple dot product? I am probably missing what `AS` is in your code.

https://github.com/meng-tang/rloss/blob/1caa759e568db2c7209ab73e73ac039ea3d7101c/pytorch/pytorch-deeplab_v3_plus/DenseCRFLoss.py#L34

Thanks a lot,
Reuben

**Comment by ReubenDo** [comment]
Does it in fact correspond to the formula in the appendix?
![image](https://user-images.githubusercontent.com/17268715/72429722-85515000-3788-11ea-9d2c-6d4706b36f5d.png)

**Comment by mkusner** [comment]
+1 I have the same question!

**Comment by mkusner** [comment]
Oh actually it looks like #3 talks about this. Is it right @meng-tang that the loss is:

![image](https://user-images.githubusercontent.com/8540881/72663326-4c63e600-39e9-11ea-90e2-740fad1f20c7.png)

Or did I miss something?

**Comment by meng-tang** [comment]
Yes, it corresponds to the formula. Sorry for inconsistent notations. A or w denotes the affinity matrix. S or X means the segmentation variable. AS is the result of the product of affinity matrix and segmentation.

Note that for dense CRF, we don't explicitly save A in our implementation. However, one can compute AS efficiently with fast Bilateral filtering.
